### DISCUSSION TITLE
Generating an assembly listing under platform.io

### DISCUSSION THREAD

**Discussion by robmaxtech** [SpenceKonde/megaTinyCore#964] [Q&A] [answered]
Hello,
I'm struggling to generate an assembly listing to look a bit in-depth what's giong on (especially what happens in a timer interrupt - the prologue and epilogue seems to be too long, way more than the numbers mentioned in Ref_Interrupts.md)
 The core offers "Sketch -> Export compiled binary" within the Arduino IDE, but how to do this in VS Code / platform.io? 
I tried
build_type = debug
build_flags = -save-temps 
but the .s files look like this:
	.ascii	"\357\375\030\026\306\300\002\35428\003\321\370\3530p\236\276"
	.ascii	"\261\232b\253\265PU7\252\036U5F\323e8\r1p\n(\370,X\265\200/."
	.ascii	"|p \022\305\2629\352\216s\245\027\256\357\024\365aV\233}a\316"
        ......

Thanks a lot for any hint!

**Comment by SpenceKonde** [comment, marked as answer]
/path/to/avr-objdump" --disassemble --source --line-numbers --demangle
--section=.text "/path/to/setch.ino.elf" > "/path/to/outputfile.lst"
(extracted from the console output of the IDE.

The temp files are unreadable because of lto - lto-supporting intermediate
files are not human readable. That's the price you pay for binaries that
are 10-20% smaller and those compile/linktime functions like
__builtin_constant_p that we make heavy use of.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by robmaxtech** [reply]
Brilliant! Works like a charm! Thanks a lot for the explanations! Your work is pure gold, really.

**Comment by SpenceKonde** [comment]
Is this something that should be added to the platformio docs?

On Fri, Jun 2, 2023 at 5:44 AM robmaxtech ***@***.***> wrote:

> Brilliant! Works like a charm! Thanks a lot for the explanations! Your
> work is pure gold, really.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/964#discussioncomment-6069913>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEW3AEZXSP7OW6YSSHWLXJGYWZANCNFSM6AAAAAAYWWOJEA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***
> com>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by robmaxtech** [reply]
I think it might be helpful! Ideally with the steps how to generate it via a custom script under platformio - this is what I tried today. Basically it shouldn't be a big deal, but I struggled with spaces in the pathname for BUILD_DIR and couldn't fix it yet. But I can contribute my findings - from the perspective of a bit naive user ;-)

**Reply by SpenceKonde** [reply]
Well, how'd you do it because I have never even run PIO?

**Reply by robmaxtech** [reply]
Ok, it work sas following (I took it from here:  https://five-embeddev.com/baremetal/platformio/  and modified it to work with spaces in pathnames):

In platformio.ini, under the [env] section, add the following two lines:

extra_scripts = post_build.py
targets = disasm


Add a new file to the project root folder (besides of platformio.ini) named 'post_build.py' (ugh, Python... indent alarm!!)
The content of this file is:

import os
Import("env")

#https://docs.platformio.org/en/latest/projectconf/advanced_scripting.html
#https://docs.platformio.org/en/latest/projectconf/section_env_build.html#projectconf-dynamic-build-flags

def after_build(source, target, env): 
    """ Run objdump on the target elf file and save the output in the top dir.
    """
    objdump=env.subst("${OBJCOPY}").replace("objcopy","objdump")
    src_elf=env.subst("\\"${BUILD_DIR}\\${PROGNAME}.elf\\"")
    #--disassembler-options=no-aliases,numeric 
    cmd=" ".join([
        objdump, "-SC","--file-start-context", "-w",
        src_elf,">","${PROGNAME}.lst"]) # 
    env.Execute(cmd)

env.AddCustomTarget(
    "disasm",
    "${BUILD_DIR}/${PROGNAME}.elf",
    after_build,
    title="Disasm 2",
    description="Generate a disassembly file on demand",
    always_build=True
)
